library/induction.red

```diff
@@ -1,85 +1,81 @@
+  elim n [
+  | zero → λ _ → refl
+  | suc n' → λ p → elim p []
+  ]
 
-def inc : nat → nat = (\ i →  suc i)
+def le/case (m : nat) : (n : nat) → (le n (suc m)) → or (path nat n (suc m)) (le n m) =
+  elim m [
+  | zero → λ n →
+    elim n [
+    | zero → λ _ → inr triv
+    | suc n' →
+      elim n' [
+      | zero → λ _ → inl refl
+      | suc _ → λ p → inr p
+      ]
+    ]
+  | suc (m' → c) →
+    elim [
+    | zero → λ _ → inr triv
+    | suc n' → λ p →
+      let r : or (path nat n' (suc m')) (le n' m') = c n' p in
+      elim r [
+      | inl p → inl (λ i -> suc (p i))
+      | inr l → inr (le/suc n' m' l)
+      ]
+    ]
+  ]
 
-def le/case (m : nat) : (n : nat) → (le n (suc m)) →  or (path nat n (suc m)) (le n m) = 
-	elim m [
-		| zero →  \ n → 
-			elim n [
-				| zero →  \ _ →  inr triv
-				| suc n' →  elim n' [
-					| zero → \ _ →  inl refl
-					| suc _ →  \ p →  inr p
-				]
-			]
-		| suc (m' →  c) →  elim [
-				| zero →  \ _ →  inr triv
-				| suc n' →  \ p →  
-				let r : or (path nat n' (suc m')) (le n' m') = c n' p in 
-					elim r [
-						| inl p →  inl (\i -> inc (p i))
-						| inr l →  inr (le/suc n' m' l)
-					]
-			]
-	]
+def weak/induction (P : nat → type) : type =
+  (P zero) → ((n : nat) → P n → P (suc n)) → ((n : nat) → P n)
 
+def complete/induction (P : nat → type) : type =
+  (P zero) → ((n : nat) → ((k : nat) → (le k n) → P k) → P (suc n)) → ((n : nat) → P n)
 
-def weak/induction (P : nat →  type)
-: type = 	(P zero) →  ((n : nat) →  P n →  P (suc n)) →  ((n : nat) → P n)  
-
-def complete/induction (P : nat →  type) : type =
-	(P zero) →  ((n : nat) →  ((k : nat) → (le k n) →  P k) → P (suc n)) →  ((n : nat) → P n)  
-
-def complete/implies/weak (P : nat →  type) : complete/induction P → weak/induction P = 
-	λ complete p0 ps →  complete p0 (\ n f →  ps n (f n (eq/implies/le n)))
-
-def weak/implies/complete : (P : nat → type) →  
-	((P' : nat →  type) →  weak/induction P') →  complete/induction P = 
-	λ P weak p0 ps →  
-		let P' : nat → type = \n → (k : nat) → (le k n) →  P k in
-		let P'0 : P' zero = \ k k/le/0 →  transport nat zero k (le/zero/implies/zero k k/le/0) P p0  in 
-		let f : (n : nat) →  (P' n) → (P' (suc n)) = 
-			\ n p'n →  \ k k/le/sn →  
-			elim
-			(le/case n k k/le/sn) [
-				| inl p →  transport nat (suc n) k (symm nat p) P (ps n p'n)
-				| inr l →  p'n k l
-			]
-		 in 
-		 let P'n : ((n : nat) → P' n) = (weak P') P'0 f in
-		 λ n →  P'n n n (eq/implies/le n)
+def complete/implies/weak (P : nat → type) : complete/induction P → weak/induction P =
+  λ complete p0 ps → complete p0 (λ n f → ps n (f n (eq/implies/le n)))
 
+def weak/implies/complete : (P : nat → type) →
+  ((P' : nat → type) → weak/induction P') → complete/induction P =
+  λ P weak p0 ps →
+    let P' : nat → type = λ n → (k : nat) → (le k n) → P k in
+    let P'0 : P' zero = λ k k/le/0 → transport nat zero k (le/zero/implies/zero k k/le/0) P p0 in
+    let f : (n : nat) → (P' n) → (P' (suc n)) =
```

Comment:
A slicker way to write this is to put `let f (n : nat) (p'n : P' n) : P' (suc n) = \ k ...`; this way we don't have to have as many lambdas.